lib/screens/customer_product_detail_page.dart

```diff
@@ -440,14 +443,47 @@ class _CustomerProductDetailPageState extends State<CustomerProductDetailPage> {
           cartItem.stockRequest?.size == _selectedItemSize);
 
       if (productIndex != -1) {
-        customer.cart.items[productIndex].stockRequest.quantity++;
+        StockItem productStock = StockItem();
+        if (_stockRequest.color != null && _stockRequest.size != null) {
+          productStock = _product.stock.items.firstWhere((stockItem) =>
+              stockItem?.color == _selectedItemColor &&
+              stockItem?.size == _selectedItemSize);
+        }
+        if (_stockRequest.color == null && _stockRequest.size != null) {
+          productStock = _product.stock.items
+              .firstWhere((stockItem) => stockItem?.size == _selectedItemSize);
+        }
+        if (_stockRequest.color != null && _stockRequest.size == null) {
+          productStock = _product.stock.items.firstWhere(
+              (stockItem) => stockItem?.color == _selectedItemColor);
+        }
+        if (customer.cart.items[productIndex].stockRequest.quantity <
+            productStock.quantity) {
+          customer.cart.items[productIndex].stockRequest.quantity++;
```

Comment:
I think this comparison

```
if (customer.cart.items[productIndex].stockRequest.quantity <
            productStock.quantity) {
```

along with the '**getting the product stock item**' can be inside the `_canAddToCart` method. This way you can minimize the edits to the rest of your code

Reply:
This change implies to make _canAddToCart async so we can use customer instance and the return of this function would be future<bool>. Do you see this okay?

Reply:
Why do you need to make it async in order to pass the customer instance?

Reply:
_updateCartModel still doing some checks if the product quantity is available or allowed, I thought that " 'getting the product stock item' can be inside the _canAddToCart method
 mean to move those validations inside _canAddToCart. Is that correct?

Reply:
Yes, I suggest something like this:

```
bool _canAddToCart(Product _product) {
    if (_product.stock.stockType == StockType.sizeAndColor) {
      return _selectedItemColor != null && _selectedItemSize != null;
    } else if (_product.stock.stockType == StockType.sizeOnly) {
      return _selectedItemSize != null;
    } else if (_product.stock.stockType == StockType.colorOnly) {
      return _selectedItemColor != null;
    }

    int productIndex = customer.cart.items.indexWhere((cartItem) =>
        cartItem.productDocumentId == widget.productDocumentId &&
        cartItem.stockRequest?.color == _selectedItemColor &&
        cartItem.stockRequest?.size == _selectedItemSize);

    bool addToCartSuccess = true;
    if (productIndex != -1) {
      StockItem productStock = getRequestedStockItem(_product, _stockRequest);
      if (customer.cart.items[productIndex].stockRequest.quantity >=
          productStock.quantity) {
        addToCartSuccess = false;
      }
    }

    return addToCartSuccess;
  }
```